src/Symfony/Component/DependencyInjection/LazyString.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection;
+
+/**
+ * A string whose value is computed lazily by a callback.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class LazyString
+{
+    private $value;
+
+    /**
+     * @param callable $callback A callable or a [Closure, method] lazy-callable
+     *
+     * @return static
+     */
+    public static function fromCallable($callback, ...$arguments): self
+    {
+        if (!\is_callable($callback) && !(\is_array($callback) && isset($callback[0]) && $callback[0] instanceof \Closure && 2 >= \count($callback))) {
+            throw new \TypeError(sprintf('Argument 1 passed to %s() must be a callable or a [Closure, method] lazy-callable, %s given.', __METHOD__, \gettype($callback)));
+        }
+
+        $lazyString = new static();
+        $lazyString->value = static function () use (&$callback, &$arguments, &$value): string {
+            if (null !== $arguments) {
+                if (!\is_callable($callback)) {
+                    $callback[0] = $callback[0]();
+                    $callback[1] = $callback[1] ?? '__invoke';
+                }
+                $value = $callback(...$arguments);
+                $callback = self::getPrettyName($callback);
+                $arguments = null;
+            }
+
+            return $value ?? '';
+        };
+
+        return $lazyString;
+    }
+
+    public function __toString()
+    {
+        if (\is_string($this->value)) {
+            return $this->value;
+        }
+
+        try {
+            return $this->value = ($this->value)();
+        } catch (\Throwable $e) {
+            if (\TypeError::class === \get_class($e) && __FILE__ === $e->getFile()) {
+                $type = explode(', ', $e->getMessage());
+                $type = substr(array_pop($type), 0, -\strlen(' returned'));
+                $r = new \ReflectionFunction($this->value);
+                $callback = $r->getStaticVariables()['callback'];
+
+                $e = new \TypeError(sprintf('Return value of %s() passed to %s::fromCallable() must be of the type string, %s returned.', $callback, static::class, $type));
+            }
+
+            if (\PHP_VERSION_ID < 70400) {
+                // leverage the ErrorHandler component with graceful fallback when it's not available
+                return trigger_error($e, E_USER_ERROR);
+            }
+
+            throw $e;
+        }
+    }
+
+    private function __construct()
```

Comment:
Useless construct ?

Reply:
it's not useless. As it is private, it forbids calling `new LazyString()` from outside the LazyString class, which enforces the usage of the static method `LazyString::fromCallable`.